src/engine.rs

```diff
@@ -288,6 +344,58 @@ impl Engine {
         sdk::sha256(&data)
     }
 
+    pub fn charge_gas_limit(
+        sender: &Address,
+        gas_limit: U256,
+        gas_price: U256,
+    ) -> Result<Wei, GasPaymentError> {
+        // Early exit as performance optimization
+        if gas_price.is_zero() {
+            return Ok(Wei::zero());
+        }
+
+        let eth_amount = Wei::new(
+            gas_limit
+                .checked_mul(gas_price)
+                .ok_or(GasPaymentError::EthAmountOverflow)?,
+        );
+        let account_balance = Self::get_balance(sender);
+        let remaining_balance = account_balance
+            .checked_sub(eth_amount)
+            .ok_or(GasPaymentError::OutOfFund)?;
+
+        Self::set_balance(sender, &remaining_balance);
+
+        Ok(eth_amount)
+    }
+
+    pub fn refund_unused_gas(
+        sender: &Address,
+        relayer: &Address,
+        prepaid_eth: Wei,
```

Comment:
This is not `prepaid_eth`, but Wei again. And also you are using other variables with `*_amount`, so I suggest having something like `prepaid_amount_for_gas`: (I'm also ok with `prepaid_amount` if you prefer a shorter name)

```suggestion
        prepaid_amount_for_gas: Wei,
```

Reply:
Done in https://github.com/aurora-is-near/aurora-engine/pull/215/commits/5bdc6d0dc4897e5892f0f72d8e1eee36602a45e0